DM-48880: avoid too-coarse spatial joins in queries

TallJimbo opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

TallJimbo committed:
Add failing test for too-coarse spatial join bug.

TallJimbo committed:
Avoid too-coarse automatic spatial joins.

A joined query term (like a dataset search or materialization) no
longer blocks the automatic addition of the most fine-grained spatial
join unless it actually has the same fine-grained relationships.

TallJimbo committed:
Downgrade warning about PostgreSQL versions to debug.

The end user can't do anything about this and usually doesn't care.

andy-slac commented on python/lsst/daf/butler/queries/overlaps.py:
```diff
@@ -259,10 +277,14 @@ def add_join_operand_connections(self, operand_dimensions: DimensionGroup) -> No
-            self._spatial_connections.merge(a_family, b_family)
-        for a_family, b_family in itertools.pairwise(operand_dimensions.temporal):
-            self._temporal_connections.merge(a_family, b_family)
+        for a_family, b_family in itertools.pairwise(getattr(operand_dimensions, kind)):
```
Can you pass `operand_dimensions.spatial` or `.temporal` into this method instead of doing `getattr`? It might help `mypy` to figure out types of those?

TallJimbo replied:
Done.  I also noticed along the way that we've been using `pairwise` here but it really should be `combinations`.

andy-slac reviewed (approved):
Looks great.

TallJimbo committed:
Fix incorrect itertools usage.

This wasn't noticed because our dimension universe only has 3.5
spatial systems (HEALpix is the 0.5).
